python/pyhail/dataset.py

```diff
@@ -266,6 +332,124 @@ def annotate_samples_list(self, input, root):
+
+        >>> conf = pyhail.TextTableConfig(impute=True)
+        >>> vds1 = (hc.read('example1.vds')
+        >>>  .annotate_samples_table('samples1.tsv', 'Sample', root='sa.phenotypes', config=conf))
+
+        This will read the file and produce annotations of the following schema::
+
+          Sample annotations:
+          sa: sa.<identifier>
+              phenotypes: sa.phenotypes.<identifier>
+                  Phenotype1: Double
+                  Phenotype2: String
+                  Age: Int
+
+        Let's import annotations from a .csv file with missing data and special characters::
+
+          $ cat samples2.tsv
+          Batch,PT-ID
+          1kg,PT-0001
+          1kg,PT-0002
+          study1,PT-0003
+          study3,PT-0003
+          .,PT-0004
+          1kg,PT-0005
+          .,PT-0006
+          1kg,PT-0007
+
+        In this case, we should:
+
+         - Escape the ``PT-ID`` column with backticks in the ``sample-expr`` argument because it contains a dash
+         - Pass the non-default delimiter ``,``
+         - Pass the non-default missing value ``.``
+         - Add the only useful column using ``code`` rather than ``root`` parameter.
+
+        >>> conf = pyhail.TextTableConfig(delimiter=',', missing='.')
+        >>> vds2 = (hc.read('example2.vds')
+        >>>  .annotate_samples_table('samples2.tsv', '`PT-ID`', code='sa.batch = table.Batch', config=conf))
+
+        Let's import annotations from a file with no header and sample IDs that need to be transformed. Suppose the vds sample IDs are of the form ``NA#####``. This file has no header line, and the sample ID is hidden in a field with other information::
+
+          $ cat samples3.tsv
+          1kg_NA12345     female
+          1kg_NA12346     male
+          1kg_NA12348     female
+          pgc_NA23456     female
+          pgc_NA23415     male
+          pgc_NA23418     male
+
+        Let's import it:
+
+        >>> conf = pyhail.TextTableConfig(noheader=True)
+        >>> vds3 = (hc.read('example3.vds')
+        >>>  .annotate_samples_table('samples3.tsv', '_0.split("_")[1]', code='sa.sex = table._1, sa.batch = table._0.split("_")[0]', config=conf))
+
+        **Common uses for the `code` argument**
+
+        Don't generate a full struct in a table with only one annotation column::
+
+          code='sa.annot = table._1'
+
+        Put annotations on the top level under `sa`::
+
+          code='sa = merge(sa, table)'
+
+        Load only specific annotations from the table::
+
+          code='sa.annotations = select(table, toKeep1, toKeep2, toKeep3)'
+
+        The above is equivalent to::
+
+          code='sa.annotations.toKeep1 = table.toKeep1,
+                sa.annotations.toKeep2 = table.toKeep2,
+                sa.annotations.toKeep3 = table.toKeep3'
+
+        **Notes**
+
+        The generality of this module allows it to load delimited text files, json, or a mixture of the two.
+
+        **Using the ``sample-expr`` argument**
```

Comment:
Fixed in several places using the approach: **Using the** ``sample-expr`` **argument**